lib/Lower/Bridge.cpp

```diff
@@ -1100,6 +1100,20 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     return Fortran::lower::FirOpBuilder::createFunction(loc, module, name, ty);
   }
 
+  /// Temporary helper to detect shapes that do not require evaluating
+  /// bound expressions at runtime or to get the shape from a descriptor.
+  static bool isConstantShape(const Fortran::semantics::ArraySpec &shape) {
+    for (const auto &susbcript : shape) {
+      const auto &lb = susbcript.lbound().GetExplicit();
+      const auto &ub = susbcript.ubound().GetExplicit();
+      if (!lb.has_value() || !Fortran::evaluate::IsConstantExpr(*lb) ||
+          !(susbcript.ubound().isAssumed() ||
+            (ub.has_value() && Fortran::evaluate::IsConstantExpr(*ub))))
```

Comment:
I'd write this as "positive logic" to make it a bit clearer to some future maintainer.  Something like:
```
if (lb && isConst(lb) && (ub && isConst(ub) || isAssumed(ub))) {
  // do nothing
} else {
  return false;
}
```

Reply:
The logic is not the most readable here indeed... rewritten in a positive way with a lambda helper. Hopefully this logic should disappear once bounds specification expr are supported.